core/drivers/cndp/cndp.cc

```diff
@@ -0,0 +1,1093 @@
+                             true);
+    pkts[i] = pkt;
+  }
+  // Free mbufs.
+  pktmbuf_free_bulk(mbufs, num_pkts_read);
+
+  return num_pkts_read;
+}
+
+uint16_t CndpPort::XskdevSendPackets(struct fwd_port *fport,
+                                     bess::Packet **pkts, int cnt) {
+  struct rte_mbuf *mbuf = (struct rte_mbuf *)(pkts[0]);
+  bool same_pool = (mbuf->pool->pool_id == fport->bess_pool->pool()->pool_id);
+  if (same_pool) {
+    // Packets belong to same memory pool.
+    return xskdev_tx_burst(fport->xsk, (void **)pkts, cnt);
+  } else {
+    LOG_FIRST_N(INFO, 1) << "Packets belong to different mempools";
+    bess::Packet *alloc_pkts[CNDP_MAX_BURST];
+    uint16_t n_pkts = fport->xsk->buf_mgmt.buf_alloc(
+        fport, (void **)alloc_pkts,
+        ((cnt < CNDP_MAX_BURST) ? cnt : CNDP_MAX_BURST));
+
+    if (n_pkts == 0) {
+      LOG_FIRST_N(WARNING, 10) << "Cannot allocate any buffers to send packets";
+      return 0;
+    }
+
+    if (n_pkts < cnt) {
+      LOG_FIRST_N(WARNING, 10)
+          << "Cannot allocate enough buffers to send all packets"
+          << "Requested=" << cnt << "Allocated=" << n_pkts;
+    }
+    // Copy packets across mempools.
+    for (uint16_t i = 0; i < n_pkts; i++) {
+      bess::utils::CopyInlined(alloc_pkts[i]->append(pkts[i]->total_len()),
+                               pkts[i]->head_data(), pkts[i]->total_len(),
+                               true);
+    }
+    // Keep below commented for debugging.
+    /*if (n_pkts > 0) {
+      LOG_FIRST_N(INFO, 1) << "Num packets to send  = " << n_pkts;
+      LOG_FIRST_N(INFO, 1)
+          << "BESS alloc pkt data len = " << alloc_pkts[0]->data_len();
+      LOG_FIRST_N(INFO, 1)
+          << "BESS alloc pkt total len = " << alloc_pkts[0]->total_len();
+      LOG_FIRST_N(INFO, 1)
+          << "BESS alloc pkt Dump(); = " << alloc_pkts[0]->Dump();
+    }*/
+    uint16_t sent = xskdev_tx_burst(fport->xsk, (void **)alloc_pkts, n_pkts);
+    if (sent < n_pkts) {
+      LOG_FIRST_N(WARNING, 10) << "Free packets which are not sent"
+                               << "Requested=" << n_pkts << " Sent=" << sent;
+      // Free allocated packets which are not sent.
+      fport->xsk->buf_mgmt.buf_free(fport, (void **)(alloc_pkts + sent),
+                                    n_pkts - sent);
+    }
+    // Free BESS packets which are sent.
+    bess::Packet::Free(pkts, sent);
+    return sent;
+  }
+
+  return 0;
+}
+
+uint16_t CndpPort::PktdevSendPackets(struct fwd_port *fport,
+                                     bess::Packet **pkts, int cnt) {
+  pktmbuf_t **mbufs = &fport->mbufs[0];
+  uint16_t n_pkts = pktdev_buf_alloc(fport->lport, mbufs, cnt);
+
+  if (n_pkts == 0) {
+    LOG_FIRST_N(WARNING, 10) << "Cannot allocate any buffers to send packets";
+    return 0;
+  }
+
+  if (n_pkts < cnt) {
+    LOG_FIRST_N(WARNING, 10)
+        << "Cannot allocate enough buffers to send all packets"
+        << "Requested=" << cnt << "Allocated=" << n_pkts;
```

Comment:
Are not spaces needed?
```suggestion
        << " Requested=" << cnt << ", Allocated=" << n_pkts;
```